crates/bevy_app/src/app.rs

```diff
@@ -754,6 +754,121 @@ impl App {
         self
     }
 
+    /// Registers the given component `R` as a [required component] for `T`.
+    ///
+    /// When `T` is added to an entity, `R` and its own required components will also be added
+    /// if `R` was not already provided. The [`Default`] `constructor` will be used for the creation of `R`.
+    ///
+    /// If a custom constructor is desired, use [`App::register_required_components_with`] instead.
+    ///
+    /// [required component]: Component#required-components
+    ///
+    /// # Panics
+    ///
+    /// Panics if `R` is already a directly required component for `T`.
+    ///
+    /// Indirect requirements through other components are allowed. In those cases, any existing requirements
+    /// will only be overwritten if the new requirement is more specific.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use bevy_app::{App, NoopPluginGroup as MinimalPlugins, Startup};
+    /// # use bevy_ecs::prelude::*;
+    /// #[derive(Component)]
+    /// struct A;
+    ///
+    /// #[derive(Component, Default, PartialEq, Eq, Debug)]
+    /// struct B(usize);
+    ///
+    /// #[derive(Component, Default, PartialEq, Eq, Debug)]
+    /// struct C(u32);
+    ///
+    /// # let mut app = App::new();
+    /// # app.add_plugins(MinimalPlugins).add_systems(Startup, setup);
+    /// // Register B as required by A and C as required by B.
+    /// app.register_required_components::<A, B>();
+    /// app.register_required_components::<B, C>();
+    ///
+    /// fn setup(mut commands: Commands) {
+    ///     // This will implicitly also insert B and C with their Default constructors.
+    ///     commands.spawn(A);
+    /// }
+    ///
+    /// fn validate(query: Query<(&A, &B, &C)>) {
+    ///     let (a, b, c) = query.single();
+    ///     assert_eq!(b, &B(0));
+    ///     assert_eq!(c, &C(0));
+    /// }
+    /// # app.update();
+    /// ```
+    pub fn register_required_components<T: Component, R: Component + Default>(
```

Comment:
Do we want `try_` and `_by_id` versions of these? Hooks have `_by_id` versions, but no `try_` variant even though they are potentially panicking.

I think a `_by_id` version wouldn't be as useful here, since the `R` generic is needed regardless, but I could see a non-panicking `try_` variant being useful.

Reply:
Yes to try from me.

Reply:
I added `try_` versions of the methods, along with an error type. This does make the methods on `App` a bit weird though, since the panicking versions return `&mut Self` like most other methods on it, while the non-panicking versions return a `Result` 🤔

I could either leave it like this, remove the return type from the panicking versions entirely, or even remove all of the registration methods from `App` and only have them on `World`, so you'd call them on `app.world_mut()`. Component hooks do the third option (`App` has no direct methods for hooks), although it is kind of nice to have them on `App` as well...

Reply:
I like the existing design best.